fix(lib): exposed testing matchers for other languages properly (#1935)
hashicorp/terraform-cdk: website/docs/cdktf/test/unit-tests.mdx

# Unit Tests

Testing your application can give you faster feedback cycles and guard you against unwanted changes. Testing is currently supported in Typescript with jest and compatible with any testing framework that supports assertions for all other languages.

> For Typescript we generate all files necessary to run jest when you run `cdktf init` so that you can start writing tests right away. If you want to add jest to an existing Typescript project, please [follow their guide](https://jestjs.io/docs/getting-started). Then, you need to add these lines in a [setup file](https://jestjs.io/docs/configuration#setupfiles-array):

Reviewer: Could you do a quick follow up and add the testing setup to all of the templates as well? Should be as simple as copying the example tests you already wrote into a file (commented out) and adding the testing helper and testing instructions

# Unit Tests

> Testing your application can give you faster feedback cycles and guard you against unwanted changes. Testing is currently supported in Typescript with jest and compatible with any testing framework that supports assertions for all other languages.

Reviewer 2: ```suggestion
Testing your application can give you faster feedback cycles and guard you against unwanted changes. Testing is supported in Typescript with Jest. For all other languages, you can use any existing testing framework that supports assertions. 
```

> We generate all files necessary to start testing when you run `cdktf init` so that you can start writing tests right away.
> 
> If you want to add jest to an existing Typescript project, please [follow their guide](https://jestjs.io/docs/getting-started). Then, you need to add these lines in a [setup file](https://jestjs.io/docs/configuration#setupfiles-array):
> 
> ```js
> const cdktf = require("cdktf");
> cdktf.Testing.setupJest();
> ```
> 
> If you want to add pytest to an existing Python project, please [follow their guide](https://docs.pytest.org/en/7.1.x/getting-started.html). The CDKTF specific assertions can be found in the cdktf package under `Testing`.
> 
> If you want to add JUnit to an existing Java project, please [follow their guide](https://maven.apache.org/plugins-archives/maven-surefire-plugin-2.12.4/examples/junit.html). The CDKTF specific assertions can be found in the cdktf package under `Testing`.
> 
> If you want to add Xunit to an existing C# project, please [follow their guide](https://xunit.net/docs/getting-started/netcore/cmdline). The CDKTF specific assertions can be found in the cdktf package under `Testing`.
> 
> If you want to add testing to an existing Go project, please [follow their guide](https://go.dev/doc/tutorial/add-a-test). The CDKTF specific assertions can be found in the cdktf package under `Testing`.

Reviewer 2: Okay how do we feel about making this an actual section? I think it's generally about how to set up testing and integrate it into your project, right? (I realize there wasn't a section before, so this is new!)

Also - when we say that we autogenerate all the files, is that true for other testing frameworks, or is it just true for Jest with TypeScript?

```suggestion
## Add Testing to Your Application

Use the following resources to add testing to your CDKTF application:
- **TypeScript:** Follow TypeScript's [Getting Started guide](https://jestjs.io/docs/getting-started) for testing with Jest. Then, add these lines in a [setup file](https://jestjs.io/docs/configuration#setupfiles-array):

    ```js
    const cdktf = require("cdktf");
    cdktf.Testing.setupJest();
    ```
- **Python:** Follow the [Get Started guide](https://docs.pytest.org/en/7.1.x/getting-started.html) for pytest. The assertions for CDKTF are available in the `cdktf` package under `Testing`.
- **Java:** Follow the [Using JUnit documentation](https://maven.apache.org/plugins-archives/maven-surefire-plugin-2.12.4/examples/junit.html). The assertions for CDKTF specific are available in the `cdktf` package under `Testing`.
- **C#:** Follow the [Getting Started with xUnit.net guide](https://xunit.net/docs/getting-started/netcore/cmdline). The assertions for CDKTF are available in the `cdktf` package under `Testing`.
- **Go:** Follow the [Add a Test guide](https://go.dev/doc/tutorial/add-a-test) in the Go documentation. The assertions for CDKTF specific are available in the `cdktf` package under `Testing`.

After you add testing to your application, run `cdktf init`. CDKTF autogenerates all of the necessary testing files.
```

Author: So it is true that all language will have all the generated files needed to start testing, great idea to add a separate section for adding testing to existing project 😁

### Write Assertions

> The Typescript example below uses `Testing.synthScope` to test a part of the application. This creates a scope to test a subset of the application and returns a JSON string representing the synthesized HCL-JSON. Then it uses custom matchers to verify the code acts as intended.

Reviewer 2: ```suggestion
The following Typescript example uses `Testing.synthScope` to test a part of the application. This creates a scope to test a subset of the application and returns a JSON string representing the synthesized HCL-JSON. Then, it uses custom matchers to verify the code acts as intended.
```

> The other examples use `Testing.synth` to test a part of the application. Given the desired scope to test, a JSON string representing the synthesized HCL-JSON is returned. Then the custom assertions under `Testing` in the cdktf package can be used to verify the code acts as intended.

Reviewer 2: ```suggestion
The examples for other languages use `Testing.synth` to test a part of the application. CDKTF returns a JSON string representing the synthesized HCL-JSON is returned. Then, the custom assertions under `Testing` in the `cdktf` package verify the code acts as intended.
```

    assert Testing.to_have_resource_with_properties(self.synthesized, Image.TF_RESOURCE_TYPE, {
            "name": "ubuntu:latest",
    }).pass_
```

```java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertTrue;
import com.hashicorp.cdktf.Testing;
import imports.docker.Container;
import imports.docker.Image;
import com.hashicorp.cdktf.MyApplicationsAbstraction; // Could be a class extending from cdktf.Resource

public class TestApplication {

  private final TerraformStack stack = new TerraformStack(Testing.app(), "stack");
  private final MyApplicationsAbstraction appAbstraction = new MyApplicationsAbstraction(stack, "resource");
  private final String synthesized = Testing.synth(stack);

  @Test
  void shouldContainContainer() {
    assertTrue(Testing.toHaveResource(synthesized, Container.TF_RESOURCE_TYPE).getPass());
  }

  @Test
  void shouldUseUbuntuImage() {
    assertTrue(Testing.toHaveResourceWithProperties(synthesized, Image.TF_RESOURCE_TYPE, new HashMap<String, Object>() {
                {
                    put("name", "ubuntu:latest");
                }
    }).getPass());
  }
}
```

```csharp
using Xunit;

> using HashiCorp.Cdktf; // MyApplicationsAbstraction - Could be a class extending from cdktf.Resource

Reviewer 3: ```suggestion
using HashiCorp.Cdktf; // MyApplicationsAbstraction - Could be a class extending from Construct
```
I think we recommend Construct now (we used to recomment extending Resource though)
What do you think @DanielMSchmidt?

Reviewer: Yeah, I agree we recommend Construct

    assert Testing.to_have_resource(self.synthesized, Container.TF_RESOURCE_TYPE).pass_

  def test_should_use_an_ubuntu_image(self):
    assert Testing.to_have_resource_with_properties(self.synthesized, Image.TF_RESOURCE_TYPE, {
            "name": "ubuntu:latest",
    }).pass_
```

```java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertTrue;
import com.hashicorp.cdktf.Testing;
import imports.docker.Container;
import imports.docker.Image;
import com.hashicorp.cdktf.MyApplicationsAbstraction; // Could be a class extending from Construct

public class TestApplication {

  private final TerraformStack stack = new TerraformStack(Testing.app(), "stack");
  private final MyApplicationsAbstraction appAbstraction = new MyApplicationsAbstraction(stack, "resource");
  private final String synthesized = Testing.synth(stack);

  @Test
  void shouldContainContainer() {
    assertTrue(Testing.toHaveResource(synthesized, Container.TF_RESOURCE_TYPE).getPass());
  }

  @Test
  void shouldUseUbuntuImage() {
    assertTrue(Testing.toHaveResourceWithProperties(synthesized, Image.TF_RESOURCE_TYPE, new HashMap<String, Object>() {
                {

>                     put("name", "ubuntu:latest");

Reviewer 4: Normally recommend against creating an anonymous map class, but otherwise need to pull in additional dependencies to keep code concise so 🤷

                }
    }).getPass());
  }
}
```

```csharp
using Xunit;
using HashiCorp.Cdktf; // MyApplicationsAbstraction - Could be a class extending from Construct
using MyCompany.MyApp;
using System;
using System.Collections.Generic;
using docker;

namespace MyCompany.MyApp{
  public class TestApplication{

>     private static TerraformStack stack = new TerraformStack(Testing.app(), "stack");

Reviewer 4: I'd opt for putting this in test setup methods rather than just static fields.

    private static MyApplicationsAbstraction appAbstraction = new MyApplicationsAbstraction(stack, "resource");
    private static string synthesized = Testing.synth(stack);

    [Fact]

>     public void shouldContainContainer(){

Reviewer 4: ```suggestion
    public void ShouldContainContainer(){
```

      Assert.True(Testing.ToHaveResource(synthesized, Container.TfResourceType).Pass);
    }

    [Fact]
    public void shouldUseUbuntuImage(){
      Assert.True(Testing.ToHaveResourceWithProperties(synthesized, Image.TfResourceType, new Dictionary<String, Object>() {
              {"name", "ubuntu:latest"}
      }).Pass);
    }
  }
}
```

```go
package main

import (
	"testing"
	"cdk.tf/go/stack/generated/kreuzwerker/docker"
	"github.com/hashicorp/terraform-cdk-go/cdktf"
	"github.com/aws/jsii-runtime-go"
)

var stack = NewMyApplicationsAbstraction(cdktf.Assertions_App(nil), "stack")
var synth = cdktf.Assertions_Synth(stack)

func TestShouldContainContainer(t *testing.T){
	assertion := cdktf.Assertions_ToHaveResource(synth, docker.Container_TfResourceType())

	if !*assertion.Pass() {
		t.Error(assertion.Message())
	}
}

func TestShouldUseUbuntuImage(t *testing.T){
	properties := map[string]interface{}{
		"name": "ubuntu:latest",
	}
	assertion := cdktf.Assertions_ToHaveResourceWithProperties(synth, docker.Image_TfResourceType(), &properties)

	if !*assertion.Pass() {
		t.Error(assertion.Message())
	}
}
```

</CodeTabs>

### Snapshot Testing

> Snapshot tests are useful when you want to make sure your infrastructure does not change unexpectedly. Snapshot Testing is currently only supported in Typescript with jest You can read more about them in the [Jest docs](https://jestjs.io/docs/snapshot-testing).

Reviewer 2: ```suggestion
Snapshot tests are useful when you want to make sure your infrastructure does not change unexpectedly. Snapshot Testing is only supported in Typescript with Jest. Refer to the [Jest docs](https://jestjs.io/docs/snapshot-testing) for details.
```

});
```

> 

Reviewer 2: We need an intro comment explaining what the examples do :)

  app_abstraction = MyApplicationsAbstraction(stack, "app-abstraction")

  def test_check_validity(self):
    assert Testing.to_be_valid_terraform(Testing.full_synth(stack)).pass_
```

```java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertTrue;
import com.hashicorp.cdktf.MyApplicationsAbstraction; // Could be a class extending from Construct

public class TestApplication {

  private final TerraformStack stack = new TerraformStack(Testing.app(), "stack");
  private final MyApplicationsAbstraction appAbstraction = new MyApplicationsAbstraction(stack, "resource");

  @Test
  void checkValidity() {
    assertTrue(Testing.toBeValidTerraform(Testing.fullSynth(stack)).getPass());
  }
}
```

```csharp
using Xunit;
using HashiCorp.Cdktf; // MyApplicationsAbstraction - Could be a class extending from Construct
using MyCompany.MyApp;
using HashiCorp.Cdktf;
using System;

namespace MyCompany.MyApp{
  public class TestApplication{

    private static TerraformStack stack = new TerraformStack(Testing.app(), "stack");
    private static MyApplicationsAbstraction appAbstraction = new MyApplicationsAbstraction(stack, "construct");

    [Fact]

>     public void checkValidity(){

Reviewer 4: ```suggestion
    public void CheckValidity(){
```
